### PR TITLE
#71 - add Beam provider

### LINKED ISSUE #71: Add Beam provider
Beam is self-called Asia Pacific’s leading micromobility company.

Cities referenece: https://partner.ridebeam.com/cities

### PR DESCRIPTION
**Pull request by BarTracz**
This should close #71

### CODE REVIEW

File: app/Importers/BeamDataImporter.php [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            $country = null;
+            foreach ($section->childNodes as $node) {
+                if ($node->nodeName === "h4") {
+                    $this->countryName = $node->nodeValue;
+                }
+                if ($node->nodeName === "div") {
+                    foreach ($node->childNodes as $div) {
+                        if ($div->nodeName === "div") {
+                            foreach ($div->childNodes as $city) {
+                                if ($city->nodeName === "img" && $city->getAttribute("src") === "https://uploads-ssl.webflow.com/63c4acbedbab5dea8b1b98cd/63d8a5b60da91e7d71298637_map-vehicle-saturn.png") {
```
Comment (MrHDOLEK) [inline app/Importers/BeamDataImporter.php:80]: img not better move this value to const ? 
same with src 
and describe these variables more to make it better to read ?

Comment (kamilpiech97) [reply app/Importers/BeamDataImporter.php:56]: yep, if it is the same string I would store him in const or smth like that.

Comment (mtracz) [reply app/Importers/BeamDataImporter.php:56]: This URL (`https://uploads-ssl.webflow.com/63c4acbedbab5dea8b1b98cd/63d8a5b60da91e7d71298637_map-vehicle-saturn.png`) should be moved to some const/variable, cause it is used twice.

This `img` I would like to left as is.

Comment (BarTracz) [reply app/Importers/BeamDataImporter.php]: changed

**Review by mtracz — APPROVED** [review]

File: app/Importers/BeamDataImporter.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    public function extract(): static
+    {
+        try {
+            $response = $this->client->get("https://partner.ridebeam.com/cities");
+            $html = $response->getBody()->getContents();
+        } catch (GuzzleException) {
+            $this->createImportInfoDetails("400", self::getProviderName());
+
+            $this->stopExecution = true;
+
```
[the remark this answers is not in the document]
Comment (BarTracz) [reply app/Importers/BeamDataImporter.php:23]: Isn`t this thing for separate issue?

[the remark this answers is not in the document]
Comment (kamilpiech97) [reply app/Importers/BeamDataImporter.php:23]: yep. let's create

**Review by kamilpiech97 — APPROVED** [review]

**Review by Letha0 — APPROVED** [review]

**Review by MrHDOLEK — APPROVED** [review]

**Review by MrHDOLEK — APPROVED** [review]

File: app/Importers/BeamDataImporter.php (hunk 1 above)
Comment (MrHDOLEK) [reply app/Importers/BeamDataImporter.php]: oki

**Review by MrHDOLEK — APPROVED** [review]

File: app/Importers/BeamDataImporter.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                        if ($div->nodeName === "div") {
+                            foreach ($div->childNodes as $city) {
+                                if ($city->nodeName === "img" && $city->getAttribute("src") === $url) {
+                                    $hasEscooters = true;
+                                } elseif ($city->nodeName === "img" && $city->getAttribute("src") !== $url) {
+                                    $hasEscooters = false;
+                                }
+
+                                if ($city->nodeName === "p" && $hasEscooters === true) {
+                                    $search = ["\u{00A0}", "\u{200D}"];
```
Comment (AleksandraKozubal) [inline app/Importers/BeamDataImporter.php:65]:
```suggestion
                                    $search = ["\u{00A0}", "\u{200D}", "Prefecture"];
```

File: app/Importers/BeamDataImporter.php [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                                    $valueToDelete = "Selangor";
+                                    $cityName = str_replace("Prefecture", "", $city->nodeValue);
```
Comment (AleksandraKozubal) [inline app/Importers/BeamDataImporter.php:67]:
```suggestion
                                    $cityName = str_replace($search, "", $city->nodeValue);
```

File: app/Importers/BeamDataImporter.php [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+                                    $cityName = preg_replace('/[\p{Hiragana}\p{Katakana}\p{Han}]+/u', "", $cityName);
+                                    $cityName = str_replace($search, "", $cityName);
```
Comment (AleksandraKozubal) [inline app/Importers/BeamDataImporter.php:69]:
```suggestion
```

**Review by AleksandraKozubal — CHANGES_REQUESTED** [review]

File: app/Importers/BeamDataImporter.php (hunk 3 above)
Comment (BarTracz) [reply app/Importers/BeamDataImporter.php]: changed

File: app/Importers/BeamDataImporter.php (hunk 4 above)
Comment (BarTracz) [reply app/Importers/BeamDataImporter.php]: changed

File: app/Importers/BeamDataImporter.php (hunk 5 above)
Comment (BarTracz) [reply app/Importers/BeamDataImporter.php:68]: changed

**Review by AleksandraKozubal — APPROVED** [review]

**Pull request merged**